src/ratelimiter.rs

```diff
@@ -0,0 +1,140 @@
+use std::time::{Duration, Instant};
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    available_bytes: u64,
+    bytes_per_second: u64,
+    last_update: Instant,
+}
+
+impl RateLimiter {
+    pub fn new(mbps_target: u64, now: Instant) -> Self {
+        let bps_target = mbps_target * 1_000_000;
+        RateLimiter {
+            available_bytes: bps_target,
+            bytes_per_second: bps_target,
+            last_update: now,
+        }
+    }
+
+    pub fn bytes_available(&self, now: Instant) -> u64 {
+        let elapsed = now - self.last_update;
+        let new_bytes = elapsed.as_secs_f32() * self.bytes_per_second as f32;
+        std::cmp::min(
+            self.available_bytes + new_bytes as u64,
+            self.bytes_per_second,
+        )
+    }
+
+    pub fn consume_bytes(&mut self, now: Instant, amount: u64) {
+        let elapsed = now - self.last_update;
+        let new_bytes = (elapsed.as_secs_f32() * self.bytes_per_second as f32) as u64;
+        self.available_bytes += new_bytes;
+        self.available_bytes = std::cmp::min(self.available_bytes, self.bytes_per_second);
+        self.available_bytes -= amount;
```

Comment:
I would either put in an assert or use `saturating_sub` (depending on whether we want to crash on or silence a bug). Now if we make a programming error and consume more than what was available, in debug mode it will crash but in release it will overflow.

Reply:
asserted!